### DISCUSSION TITLE
How to get 'gx' to do xdg-open while ntree-vim disables netrw

### DISCUSSION THREAD

**Discussion by ikke-t** [nvim-tree/nvim-tree.lua#2714] [Q&A] [answered]
Hi,

netrw offers gx command, which runs system open. But nvim-tree instructions guide to disable netrw, after which the 'gx' command to open urls under cursor disappears too.

How to map back gx? I tried this, but it is not correct syntax. I don't know lua, so how could I get something like this to work? 

```
  { -- nvim-tree.lua adds the file browser
    'nvim-tree/nvim-tree.lua',
    event = 'VimEnter',
    opts = {
      system_open = { cmd = 'flatpak-xdg-open', args = {} },
    },
    dependencies = {
      { 'nvim-tree/nvim-web-devicons' },
    },
    keys = {
      -- FIX: https://www.reddit.com/r/neovim/comments/ro6oye/open_link_from_neovim/
      {
        'gx',
        mode = { 'n' },
        vim.fn.jobstart('flaptpak-xdg-open', vim.fn.expand '<cword>'),
        desc = '[X]dg Open',
      },
    },
  },
```

The above fails with some syntax error:
```
Error detected while processing /var/home/itengval/.config/nvim/init.lua:
Failed to load `custom.plugins.plugins`

Vim:E475: Invalid argument: expected dictionary

# stacktrace:
  - lua/custom/plugins/plugins.lua:80
  - init.lua:145
Press ENTER or type command to continue

```

If you question why I need flatpak-xdg-open instead of xdg-open, I need flatpak-xdg-open, as I'm in container while running this on Fedora Silverblue...

**Comment by ikke-t** [comment]
went around it [in init.lua](https://github.com/ikke-t/nvim-config/blob/d76e45b279f281706d2ad41d8b9989f8c272dd2a/init.lua#L13) until I get something better.

**Comment by alex-courtis** [comment]
I might be misunderstanding; can you simply use "Run System", default `s` ?

Edit: Change `system_open.cmd` to `flatpak-xdg-open` or similar.

See `:help nvim-tree-api.node.run.system()` and `:help nvim-tree.system_open.cmd`

**Comment by ikke-t** [comment]
Thanks, perhaps that's about the same as I ended up having there (vim.fn.system), or does it do something different?

```
      {
        -- as we don't have netew, we loose gx. Add it back:
        -- https://www.reddit.com/r/neovim/comments/11zgf64/gxnvim_open_links_and_more_without_netrw/
        -- will change in nvim 0.10.0 to vim.ui.open
        'gx',
        mode = { 'n' },
        function()
          local cmd
          local url = vim.fn.expand '<cfile>'
          local is_container = vim.fn.filereadable '/run/.containerenv'
          if is_container then
            cmd = 'flatpak-xdg-open'
          else
            cmd = 'xdg-open '
          end
          if string.match(url, 'https') == 'https' or string.match(url, 'http') == 'http' then
            vim.fn.system { cmd, url }
          else
            return print '💁 Woops no url under cursor 🙅'
          end
        end,
        desc = '[X]dg Open current url',
      },
```

Just for clarity, this is not in nvim-tree panel, but anywhere in any buffer that has url under cursor. I will remove the html check from above, as then it will just open any application the uri gives.

**Comment by ikke-t** [comment]
My lack of lua understanding makes me fail here. How can you call the function in keys{}? It always says api is undefined global or smth. If I add function there, it doesn't complain when there is require, but it won't do anything either.

```
  { -- nvim-tree.lua adds the file browser
    'nvim-tree/nvim-tree.lua',
    event = 'VimEnter',
    opts = {
      system_open = { cmd = 'xdg-open', args = {} },
    },
    keys = {
      {
        -- as we don't have netew, we loose gx. Add it back:
        -- https://www.reddit.com/r/neovim/comments/11zgf64/gxnvim_open_links_and_more_without_netrw/
        -- will change in nvim 0.10.0 to vim.ui.open
        'gx',
        mode = { 'n' },
        function()
          require('nvim-tree.api').node.run.system(vim.fn.expand '<cfile>')
         --  nvim-tree-api.node.run.system(vim.fn.expand '<cfile>')
        end,
        -- require('nvim-tree.api').nvim-tree-api.node.run.system(vim.fn.expand '<cfile>'),
        -- require('nvim-tree.api').node.run.system(vim.fn.expand '<cfile>'),
         -- nvim-tree-api.node.run.system(vim.fn.expand '<cfile>'),
        require('nvim-tree.api').node.run.system(vim.fn.expand '<cfile>'),
        -- function()
        --   local cmd
        --   local url = vim.fn.expand '<cfile>'
        --   local is_container = vim.fn.filereadable '/run/.containerenv'
        --   if is_container then
        --     cmd = 'flatpak-xdg-open'
        --   else
        --     cmd = 'xdg-open '
        --   end
        --   if string.match(url, 'https') == 'https' or string.match(url, 'http') == 'http' then
        --     vim.fn.system { cmd, url }
        --   else
        --     return print '💁 Woops no url under cursor 🙅'
        --   end
        -- end,
        desc = '[X]dg Open current url',
      },
    },
  },

```

**Comment by ikke-t** [comment, marked as answer]
This is the simplest I get it into, and I link the xdg-open to flatpak-xdg-open in my .local/bin:
```
  { -- nvim-tree.lua adds the file browser
    'nvim-tree/nvim-tree.lua',
    opts = {
      system_open = { cmd = 'xdg-open', args = {} },
    },
    keys = {
      {
        -- as we don't have netrw, we loose gx. Add it back:
        -- https://www.reddit.com/r/neovim/comments/11zgf64/gxnvim_open_links_and_more_without_netrw/
        -- will change in nvim 0.10.0 to vim.ui.open
        'gx',
        mode = { 'n' },
        function()
          vim.fn.system { 'xdg-open', vim.fn.expand '<cfile>' }
        end,
        desc = '[X]dg Open current url',
      },
    },
  },
```

**Reply by alex-courtis** [reply]
I'm still unclear as to why this custom function is necessary. You can map `gx` to `s` (Run System) in your `on_attach`:

```lua
-- unmap default Run System (optional)
vim.keymap.del('n', 's', { buffer = bufnr })

-- map gx to Run System
vim.keymap.set('n', 'gx',       api.node.run.system,                 opts('Run System'))
```

That will use the `system_open = { cmd = 'xdg-open', args = {} }` opts you specified.

See `:h nvim-tree-quickstart-custom-mappings` and `:h nvim-tree-mappings`

**Reply by alex-courtis** [reply]
You can set your alternative system open via `system_open = { cmd = 'flatpak-xdg-open', }` so you don't need a link.

**Reply by ikke-t** [reply]
I believe we talk about different things. I think you talk about case when you sit in nvim-tree side menu. Where as I am sitting inside the buffer that has any file open, and cursor is blinkin on url inside the file.

**Reply by gegoune** [reply]
You might want to look at `:h vim.ui.open()`, and particularly: https://github.com/neovim/neovim/blob/cff8c15977bc5b012b64ffa8a802b0b0b1c2a87c/runtime/lua/vim/_defaults.lua#L68

Also I wouldn't link this mapping to nvim-tree at all as they really don't have anything in common.

**Reply by ikke-t** [reply]
Thanks. Could it be that the vim.ui.open is not yet in my 0.9.5 nvim (fedora 40)? It won't work nor can I find it from :help. Someone mentioned it comes only in later releases.

I have it mapped while nvim-tree loads, as 'gx' functionality is lost while using it. Nvim-tree instructions tells I need to remove netrw (which provides 'gx') while installing nvim-tree, so I added the corresponding keybinding here to remember to remove it if I would not have nvim-tree. So in that sense it's linked, at least in my head.

**Reply by gegoune** [reply]
It might be on nightly only then. 🤷
